File: packages/ts/src/core/storage/credentialStore.ts
/**
 * Pluggable credential storage facade.
 *
 * The SDK persists the device id minted at first activation and the license
 * key issued by `/v1/licenses/activate` so subsequent process boots do not
 * have to re-bootstrap the user. The storage layer is a facade so the same
 * `Isa.withLicense({...})` call works in three runtimes:
 *
 * 1. React Native — caller passes an AsyncStorage-shaped adapter.
 * 2. Browser     — caller passes a localStorage-shaped adapter.
 * 3. Node        — the SDK ships an in-memory default; a Node fs adapter
 *                  can be supplied where persistence across process boots
 *                  is required.
 *
 * The interface is intentionally tiny (get / set / remove) so any KV-shaped
 * storage can be adapted in a few lines. All values are strings; callers
 * MUST NOT pass JSON to `set` — the SDK serializes once via the key naming
 * convention defined in `CREDENTIAL_KEYS`.
 *
 * Operations may be sync or async; the facade always returns Promises.
 */

/** Persistent key/value store for SDK credentials. */
export interface CredentialStore {
  /** Read a value by key. Returns `undefined` when absent. */
  get(key: string): Promise<string | undefined>;
  /** Write a value. Overwrites any prior value. */
  set(key: string, value: string): Promise<void>;
  /** Remove a value. No-op if absent. */
  remove(key: string): Promise<void>;
}

/** Canonical key names the SDK uses inside any {@link CredentialStore}. */
export const CREDENTIAL_KEYS = {
  /** Per-device identifier; HMAC signing key for `X-Device-Signature`. */
  deviceId: 'isa.deviceId',
  /** License key minted at `/v1/licenses/activate`. */
  licenseKey: 'isa.licenseKey',
} as const;

/**
 * In-memory {@link CredentialStore}. Default when the caller supplies no
 * persistent adapter. State survives the process but NOT a restart — for
 * cross-boot persistence, plug in an AsyncStorage / localStorage / fs
 * adapter via `Isa.withLicense({ credentialStore: ... })`.
 */
export function inMemoryCredentialStore(): CredentialStore {
  const state = new Map<string, string>();
  return {
    async get(key) {
      return state.get(key);
    },
    async set(key, value) {
      state.set(key, value);
    },
    async remove(key) {
      state.delete(key);
    },
  };
}

/**
 * Adapt an AsyncStorage-shaped object (React Native, MMKV, expo-secure-store)
 * into a {@link CredentialStore}. The duck-typed interface matches both
 * `@react-native-async-storage/async-storage` and `localStorage` (after
 * wrapping in async — see {@link fromLocalStorage}).
 */
export interface AsyncStorageLike {
  getItem(key: string): Promise<string | null> | string | null;
  setItem(key: string, value: string): Promise<void> | void;
  removeItem(key: string): Promise<void> | void;
}

/** Wrap an {@link AsyncStorageLike} into the SDK's {@link CredentialStore}. */
export function fromAsyncStorage(storage: AsyncStorageLike): CredentialStore {
  return {
    async get(key) {
      const value = await Promise.resolve(storage.getItem(key));
      return value == null ? undefined : value;
    },
    async set(key, value) {
      await Promise.resolve(storage.setItem(key, value));
    },
    async remove(key) {
      await Promise.resolve(storage.removeItem(key));
    },
  };
}

/**
 * Adapt a synchronous `Storage`-shaped object (browser `localStorage` /
 * `sessionStorage`) into the SDK's async {@link CredentialStore}.
 */
export interface SyncStorageLike {
  getItem(key: string): string | null;
  setItem(key: string, value: string): void;
  removeItem(key: string): void;
}

/** Wrap a synchronous `Storage` (e.g., `localStorage`) into a {@link CredentialStore}. */
export function fromLocalStorage(storage: SyncStorageLike): CredentialStore {
  return {
    async get(key) {
      const value = storage.getItem(key);
      return value == null ? undefined : value;
    },
    async set(key, value) {
      storage.setItem(key, value);
    },
    async remove(key) {
      storage.removeItem(key);
    },
  };
}

File: packages/python/src/sah_sdk/core/credential_store.py
"""Pluggable credential storage facade.

Python mirror of ``packages/ts/src/core/storage/credentialStore.ts``. The
SDK persists the device id minted at first activation and the license key
issued by ``/v1/licenses/activate`` so subsequent process boots do not
have to re-bootstrap the user. The storage layer is a facade so the same
``Isa.with_license(...)`` call works in three runtimes:

1. Server (default) — the SDK ships an in-memory store.
2. Disk — :func:`from_path_dict` adapter for a JSON-on-disk dict (caller
   supplies the read/write functions).
3. Custom — any object implementing the :class:`CredentialStore` protocol.

The interface is intentionally tiny (get / set / remove) so any KV-shaped
storage can be adapted in a few lines. All values are strings.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Final, Protocol


class CredentialStore(Protocol):
    """Persistent key/value store for SDK credentials."""

    def get(self, key: str) -> str | None: ...
    def set(self, key: str, value: str) -> None: ...
    def remove(self, key: str) -> None: ...


class CredentialKeys:
    """Canonical key names the SDK uses inside any :class:`CredentialStore`."""

    DEVICE_ID: Final[str] = "isa.deviceId"
    LICENSE_KEY: Final[str] = "isa.licenseKey"
    ORDER_ID: Final[str] = "isa.orderId"


# Backward-compatible module-level alias mirroring the TS ``CREDENTIAL_KEYS``.
CREDENTIAL_KEYS: Final[CredentialKeys] = CredentialKeys()


class InMemoryCredentialStore:
    """In-memory credential store. Default when no adapter is supplied.

    State survives the process but NOT a restart — for cross-boot
    persistence, plug in a custom :class:`CredentialStore` via
    ``Isa.with_license(credential_store=...)``.
    """

    __slots__ = ("_data",)

    def __init__(self) -> None:
        self._data: dict[str, str] = {}

    def get(self, key: str) -> str | None:
        return self._data.get(key)

    def set(self, key: str, value: str) -> None:
        self._data[key] = value

    def remove(self, key: str) -> None:
        self._data.pop(key, None)


def in_memory_credential_store() -> CredentialStore:
    """Construct a fresh in-memory store."""
    return InMemoryCredentialStore()


def from_mapping(
    *,
    getter: Callable[[str], str | None],
    setter: Callable[[str, str], None],
    remover: Callable[[str], None],
) -> CredentialStore:
    """Adapt arbitrary getter/setter/remover callables into a CredentialStore.

    Useful for wiring a custom backend (e.g. a database, a keychain, a
    cloud secret store) without subclassing.
    """

    class _Adapter:
        __slots__ = ()

        def get(self, key: str) -> str | None:
            return getter(key)

        def set(self, key: str, value: str) -> None:
            setter(key, value)

        def remove(self, key: str) -> None:
            remover(key)

    return _Adapter()
